// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: exercise/v1/create_custom_exercise.proto

package exercisev1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// define the regex for a UUID once up-front
var _create_custom_exercise_uuidPattern = regexp.MustCompile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

// Validate checks the field values on CreateCustomExerciseRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateCustomExerciseRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateCustomExerciseRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateCustomExerciseRequestMultiError, or nil if none found.
func (m *CreateCustomExerciseRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateCustomExerciseRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if err := m._validateUuid(m.GetUserId()); err != nil {
		err = CreateCustomExerciseRequestValidationError{
			field:  "UserId",
			reason: "value must be a valid UUID",
			cause:  err,
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := CreateCustomExerciseRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return CreateCustomExerciseRequestMultiError(errors)
	}

	return nil
}

func (m *CreateCustomExerciseRequest) _validateUuid(uuid string) error {
	if matched := _create_custom_exercise_uuidPattern.MatchString(uuid); !matched {
		return errors.New("invalid uuid format")
	}

	return nil
}

// CreateCustomExerciseRequestMultiError is an error wrapping multiple
// validation errors returned by CreateCustomExerciseRequest.ValidateAll() if
// the designated constraints aren't met.
type CreateCustomExerciseRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateCustomExerciseRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateCustomExerciseRequestMultiError) AllErrors() []error { return m }

// CreateCustomExerciseRequestValidationError is the validation error returned
// by CreateCustomExerciseRequest.Validate if the designated constraints
// aren't met.
type CreateCustomExerciseRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateCustomExerciseRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateCustomExerciseRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateCustomExerciseRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateCustomExerciseRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateCustomExerciseRequestValidationError) ErrorName() string {
	return "CreateCustomExerciseRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateCustomExerciseRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateCustomExerciseRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateCustomExerciseRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateCustomExerciseRequestValidationError{}

// Validate checks the field values on CreateCustomExerciseResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateCustomExerciseResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateCustomExerciseResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateCustomExerciseResponseMultiError, or nil if none found.
func (m *CreateCustomExerciseResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateCustomExerciseResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ExerciseId

	if len(errors) > 0 {
		return CreateCustomExerciseResponseMultiError(errors)
	}

	return nil
}

// CreateCustomExerciseResponseMultiError is an error wrapping multiple
// validation errors returned by CreateCustomExerciseResponse.ValidateAll() if
// the designated constraints aren't met.
type CreateCustomExerciseResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateCustomExerciseResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateCustomExerciseResponseMultiError) AllErrors() []error { return m }

// CreateCustomExerciseResponseValidationError is the validation error returned
// by CreateCustomExerciseResponse.Validate if the designated constraints
// aren't met.
type CreateCustomExerciseResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateCustomExerciseResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateCustomExerciseResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateCustomExerciseResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateCustomExerciseResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateCustomExerciseResponseValidationError) ErrorName() string {
	return "CreateCustomExerciseResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateCustomExerciseResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateCustomExerciseResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateCustomExerciseResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateCustomExerciseResponseValidationError{}
